source/extensions/quic/connection_id_generator/envoy_deterministic_connection_id_generator.cc

```diff
@@ -0,0 +1,94 @@
+#include "source/extensions/quic/connection_id_generator/envoy_deterministic_connection_id_generator.h"
+
+#include "source/common/network/socket_option_impl.h"
+#include "source/common/quic/envoy_quic_utils.h"
+
+#include "quiche/quic/load_balancer/load_balancer_encoder.h"
+
+namespace Envoy {
+namespace Quic {
+
+absl::optional<quic::QuicConnectionId>
+EnvoyDeterministicConnectionIdGenerator::GenerateNextConnectionId(
+    const quic::QuicConnectionId& original) {
+  auto new_cid = DeterministicConnectionIdGenerator::GenerateNextConnectionId(original);
+  if (new_cid.has_value()) {
+    adjustNewConnectionIdForRoutine(new_cid.value(), original);
+  }
+  return new_cid;
+}
+
+absl::optional<quic::QuicConnectionId>
+EnvoyDeterministicConnectionIdGenerator::MaybeReplaceConnectionId(
+    const quic::QuicConnectionId& original, const quic::ParsedQuicVersion& version) {
+  auto new_cid = DeterministicConnectionIdGenerator::MaybeReplaceConnectionId(original, version);
+  if (new_cid.has_value()) {
+    adjustNewConnectionIdForRoutine(new_cid.value(), original);
+  }
+  return new_cid;
```

Comment:
This should check if new_cid == original after all the changes and return nullopt if so.